packages/pluggable-widgets-tools/src/typings-generator/generateClientTypes.ts

```diff
@@ -138,8 +138,63 @@ function toClientPropType(
             if (!prop.returnType || prop.returnType.length === 0) {
                 throw new Error("[XML] Expression property requires returnType element");
             }
-            const type = toAttributeClientType(prop.returnType[0].$.type);
-            return prop.$.dataSource ? `ListExpressionValue<${type}>` : `DynamicValue<${type}>`;
+
+            const returnTypeProp = prop.returnType[0].$;
+            const assignableTo = returnTypeProp.assignableTo;
+            const type = returnTypeProp.type;
+
+            if ((type && assignableTo) || (!type && !assignableTo)) {
+                throw new Error(
+                    "[XML] Invalid return type for expression property: either type or assignableTo must be specified."
+                );
+            }
+
+            if (assignableTo) {
+                const resolvedProperty = resolveProp(assignableTo);
+                if (!resolvedProperty) {
+                    throw new Error(
+                        `[XML] Invalid return type for expression property: invalid property path '${assignableTo}' in assignableTo attribute.`
+                    );
+                } else if (resolvedProperty.$.type != "attribute") {
+                    throw new Error(
+                        `[XML] Invalid return type for expression property: assignableTo property '${assignableTo}' must be of type Attribute.`
+                    );
+                } else {
+                    const allowedAssignableToTypes = [
```

Comment:
```suggestion
                    const allowedTypes = [
```
Should be clear enough in context. (And, together with one fewer level of indentation, will hopefully allow Prettier to put this array on one line? It takes up a lot of space 😛)